src/background/config_backend.js

```diff
@@ -0,0 +1,87 @@
+import Logger from "../logger.js";
+import DBUtils from "../utilities.js";
+
+const defaultConfig = {
+  displayWaveform: false
+};
+
+export default class ConfigBackend {
+  constructor() {
+    this.log = new Logger();
+    this.dbUtils = new DBUtils();
+
+    this.boundConnectionListenerCallback = ConfigBackend.connectionListenerCallback.bind(
+      this
+    );
+
+    this.boundPortListenerCallback = ConfigBackend.portListenerCallback.bind(
+      this
+    );
+  }
+
+  init() {
+    this.log.info("initializing ConfigBackend");
+
+    chrome.runtime.onConnect.addListener(this.boundConnectionListenerCallback);
+  }
+
+  static connectionListenerCallback(port) {
+    this.log.info("connection listener callback");
+    if (port.name !== "bandcamplabelview") {
+      this.log.error(
+        `Unexpected chrome.runtime.onConnect port name: ${port.name}`
+      );
+      return;
+    }
+
+    this.port = port;
+    this.port.onMessage.addListener(this.boundPortListenerCallback);
+  }
+
+  static async portListenerCallback(msg) {
+    this.log.info("port listener callback");
+
+    const db = await this.dbUtils.getDB("config");
+
+    this.setupDB(db);
+
+    if (msg.config) this.synchronizeConfig(db, msg.config); //todo: remove this, it's not used
+
+    if (msg.toggleWaveformDisplay) this.toggleWaveformDisplay(db);
+
+    if (msg.requestConfig) this.broadcastConfig(db);
+  }
+
+  async synchronizeConfig(db, config) {
+    let db_config = await db.get("config", "config");
+    const merged_config = ConfigBackend.mergeData(db_config, config);
+
+    await db.put("config", merged_config, "config");
+    this.port.postMessage({ config: merged_config });
+  }
+
+  async toggleWaveformDisplay(db) {
+    this.log.info("toggleing waveform display");
+
+    let db_config = await db.get("config", "config");
```

Comment:
Out of curiosity, have you done any profiling on these transactions? I'm curious how expensive awaiting potentially-blocking calls is for a UI update.

Reply:
@emcniece is it actually possible to block? 
The frontend doesn't sit blocked waiting for this to complete. This happens in the background and then when it's done, it triggers an update in the frontend through the `port.onMessage` callback...

unless I am missing something...

Reply:
`await` is a way to halt flow until async methods return, so I'm assuming that `db.get()` is async and potentially blocking. Since it's all internal to the browser perhaps it's so quick that the thread blocking isn't noticeable. Might be exacerbated by slow processors but nobody's using this on mobile :)

Generally it's good practice to decouple UI changes from potentially long-running state transactions but this likely won't be noticeable.

Reply:
@emcniece ah, now I understand the concern. Not that the plugin could stall but that the UI change could just take forever to show up?

Reply:
@emcniece this happens sometimes with the preview markers. It is an acceptable risk for not having to manage state.